man/findcon.1
.TH findcon 1
.SH NAME
findcon \- SELinux file context search tool
.SH SYNOPSIS
.B findcon
FCLIST [OPTIONS] [EXPRESSION]
.SH DESCRIPTION
.PP
.B findcon
allows the user to search for files with a specified context.
Results can be filtered by object class as described below.
.SH FCLIST
The
.B findcon
tool operates upon a file context list source.  There are three valid
file context lists.
.IP directory
If
.B FCLIST
is a name of a directory then begin the search at that directory and
recurse within it.  Be sure there are no circular mounts within it.
.IP file_contexts
If
.B FCLIST
is the name of a file_contexts file (e.g.,
/etc/selinux/strict/contexts/files/file_contexts) then open that file
and find matching entries.
.IP database
If
.B FCLIST
is the name of a database as created by a previous run of
.B indexcon
or
.B apol
then open the database and execute queries into it.
.SH EXPRESSION
.P
The following options allow the user to specify which files to print.
A file must meet all specified criteria.
If no expression is provided, all files are printed.
.IP "-t TYPE, --type=TYPE"
Search for files with a context containing the type TYPE.
.IP "-u USER, --user=USER"
Search for files with a context containing the user USER.
.IP "-r ROLE, --role=ROLE"
Search for files with a context containing the role ROLE.
.IP "-m RANGE, --mls-range=RANGE"
Search for files with a context with the MLS range of RANGE.  Note
that
.B findcon
ignores the SELinux translation library, if present.  In addition,
this flag is ignored if the
.B FCLIST
has no MLS information.
.IP "--context=CONTEXT"
Search for files matching this partial context.  This flag overrides
-t, -u, -r, and -m.
.IP "-p PATH, --path=PATH"
Search for files which include PATH.
.IP "-c CLASS, --class=CLASS"
Search only files of object class CLASS.
.SH OPTIONS
The following additional options are available.
.IP "-R, --regex"
Search using regular expressions instead of exact string matching.
This option does not affect the --class flag.
.IP "-h, --help"
Print help information and exit.
.IP "-V, --version"
Print version information and exit.
.SH PARTIAL CONTEXT
The
.B --context
flag specifies a partial context, which is a a colon separated list of
user, role, and type.  If the system supports MLS, the context may
have a fourth field that gives the range.  If a field is not specified
or is the literal asterisk, then the query will always match the field.
.SH OBJECT CLASSES
Valid object class strings are
.PP
block,
char,
dir,
fifo,
file,
link, or
sock.
.SH NOTE
The findcon utility always operates on "raw" SELinux file contexts.
If the system has an installed translation library (i.e., libsetrans),
those translations are ignored in favor of reading the original
contexts from the filesystem (if FCFILE is a directory).
.SH EXAMPLES
.TP
.B findcon .
Find every context in the current directory and all of its
subdirectories.
.TP
.B findcon -u user_u .
Find every context whose user is user_u in the current directory and
all subdirectories.
.TP
.B findcon -u system_u -t bin_t file_contexts
Find entries user system_u and type bin_t within a file_contexts file,
assuming that file_contexts is a file contexts file.
.TP
.B findcon --context=system_u::bin_t file_contexts
This is equivalent to the previous example.
.TP
.B findcon --context=system_u:*:bin_t:* file_contexts
This is also equivalent to the above example.
.SH AUTHOR
This manual page was written by Jeremy A. Mowery <jmowery@tresys.com>.
.SH COPYRIGHT
Copyright(C) 2003-2007 Tresys Technology, LLC
.SH BUGS
Please report bugs via an email to setools-bugs@tresys.com.
.SH SEE ALSO
replcon(1), indexcon(1)


NEWS
SETools 3.3:

SETools:

  * SETools now has an external dependency upon libsqlite3 >= 3.2.  The
    supplied configure script will enforce this dependency.

  * pkg-config scripts are installed with the SETools libraries.

libsefs:

  * Rewrite of library to have proper namespaces and much more usable
    object-oriented design.

  * SWIG wrappers generated for this library if the appropriate
    configure flags are set.

findcon, searchcon:

  * Merge searchcon's functionality into findcon.  The searchcon tool
    has been removed from SETools.

indexcon, replcon:

  * Updated to use new libsefs design.

apol:

  * Updated to use new libsefs design.

  * Modified to use the SWIG Tcl interface rather than a custom C
    library.  apol is now a combination of a Tcl script (simply called
    'apol') and associated packages that are required at runtime.

  * Neverallow rules are only loaded and expanded when the user
    performs a search for them.  This will dramatically speed up
    initial policy load time.

awish:

  * awish is no longer needed and thus has been removed from SETools.

sediff, sediffx:

  * Instead of differentiating "AV rules" or "TE rules", user now
    specifies which particular rule to compare (allow, dontaudit,
    type_transition, etc.).

  * Neverallow rules are only loaded and expanded when the user
    performs a diff upon them.  This will dramatically speed up
    initial policy load time.

==================================================

SETools 3.2:

libapol, libqpol, libesaudit, libpoldiff:

  * If --enable-swig-python is given during configure time, the build
    system will create Python SWIG wrappers for these libraries.

  * If --enable-swig-java is given during configure time, the build
    system will create Java SWIG wrappers for these libraries.

libpoldiff:

  * Provides ability to diff levels, categories, and range_transition
    statements.  Provides ability to diff modifications to a user's
    default level and permitted MLS range, assuming the policies are MLS.
    
apol, seaudit, sediffx:

  * Introduces 'policy list', a small text file that contains
    references to a base policy and any number of modules.  After
    selecting the base policy and modules in a tool's open policy
    dialog, click on 'Export' to write a policy list to disk.  That
    policy list then may be imported into the same tool or any other
    graphical SETools application.  This file also may be specified
    on the command line for all tools that load a policy.

sesearch, apol:

  * Provides full support for version 21 policy (i.e., object classes for
    range_transition statements).

sediff:

  * Shows MLS diffs.

sediffx:

  * Shows MLS diffs.

  * Provides option to show line numbers that contributed just to a
    specific AV rule's permission by clicking that AV rule's permissions.
    This is in addition to existing sediffx behavior that showed all lines
    associated with a particular rule difference.

  * Provides support for type joins and splits within the type remap dialog.

==================================================

SETools 3.1:

SETools:

  * All tools that open a policy now support loadable policy modules.
    Command line tools expect the first module to be a base module
    followed optionally by any other modules.  Graphical tools have
    a new open policy dialog to select a base module and any number of
    additional modules.

  * Release of RPM packages that are compatible with Fedora Core 5 and
    6.  The spec and support files are in packages/rpm.

libapol:

  * New class apol_policy_path_t to represent a base policy and any
    number of modules.  Use this whenever referring to the file or
    files constituting a policy.
  
libqpol:

  * Policy features such as attribute names or MLS can now be queried
    individally via qpol_policy_has_capability() rather than inferred
    by policy type and version.

  * New class qpol_module_t to represent a particular policy module
    prior to it being linked into a base policy (qpol_policy_t).

libseaudit:

  * Rewrite of library to have proper namespaces.  libseaudit is now
    fully documented and suitable for third-party users.

seaudit:

  * Rewrite to use new libseaudit.

  * Numerous tweaks to the interface to be more user friendly.

seaudit-report:

  * Rewrite to use new libseaudit.

sediffx:

  * Numerous tweaks to the interface to be more user friendly.

==================================================

SETools 3.0.1

SETools:

  * All code has been indented uniformly via the 'make indent' target.

SETools libraries:

  * All libraries now have a get_version() function.

libqpol:

  * Syntactic rule table is now off by default; it requires an
    explicit call to qpol_policy_build_syn_rule_table() to create it.

libapol:

  * apol_policy_t is now an opaque structure.  apol_permmap_t is no
    longer a public declaration.

  * avrule and terule queries now have full "syntactic" searching
    features.

apol & sesearch:

  * If loaded policy is source, the new syntactic search algorithm is used.

==================================================

SETools 3.0

SETools:

  * Moved entire project to autoconf build system.  This will detect
    dependencies correctly and make it easier to integrate with Linux
    distributions and their packaging systems.

SETools libraries:

  * Rewrite of back-end of SETools to use libsepol data structures.
    Most things should run notably faster.

  * All exported library functions have a standardized naming
    convention and are fully documented.  This will prove helpful for
    third-party developers integrating SETools into their own
    projects.

  * For policies version 15 or greater, domain transition analysis now
    takes into consideration setexec permission and/or type_transition
    rules.

apol:

  * Rework Tk interface to fit on 1024x768 displays.

  * Indirect matching of attributes now works with rule searches.

sediff:


secmds/findcon.cc
/**
 * @file
 *
 * Search a fclist (either the disk, a database generated by indexcon
 * or apol, or a file_contexts file for entries that match a given
 * SELinux file context.
 *
 *  @author Jeremy A. Mowery jmowery@tresys.com
 *  @author Jason Tang jtang@tresys.com
 *
 *  Copyright (C) 2003-2007 Tresys Technology, LLC
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <config.h>

#include <sefs/db.hh>
#include <sefs/fcfile.hh>
#include <sefs/filesystem.hh>
#include <sefs/entry.hh>
#include <sefs/query.hh>

using namespace std;

#include <errno.h>
#include <iostream>
#include <getopt.h>
#include <string>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#define COPYRIGHT_INFO "Copyright (C) 2003-2007 Tresys Technology, LLC"

static struct option const longopts[] = {
	{"class", required_argument, NULL, 'c'},
	{"type", required_argument, NULL, 't'},
	{"user", required_argument, NULL, 'u'},
	{"role", required_argument, NULL, 'r'},
	{"mls-range", required_argument, NULL, 'm'},
	{"path", required_argument, NULL, 'p'},
	{"regex", no_argument, NULL, 'R'},
	{"help", no_argument, NULL, 'h'},
	{"version", no_argument, NULL, 'V'},
	{NULL, 0, NULL, 0}
};

static void usage(const char *program_name, bool brief)
{
	cout << "Usage: " << program_name << " FCLIST [OPTIONS] [EXPRESSION]" << endl << endl;
	if (brief)
	{
		cout << "\tTry " << program_name << " --help for more help." << endl << endl;
		return;
	}

	cout << "Find files matching the given SELinux context." << endl << endl;

	cout << "FCLIST is a directory name, a file_contexts file, or a database" << endl;
	cout << "created by a previous run of indexcon." << endl;
	cout << endl;

	cout << "EXPRESSION:" << endl;
	cout << "  -t TYPE,  --type=TYPE          find contexts with type TYPE" << endl;
	cout << "  -u USER,  --user=USER          find contexts with user USER" << endl;
	cout << "  -r ROLE,  --role=ROLE          find contexts with role ROLE" << endl;
	cout << "  -m RANGE, --mls-range=RANGE    find contexts with MLS range RANGE" << endl;
	cout << "  -p PATH,  --path=PATH          find files in PATH" << endl;
	cout << "  -c CLASS, --class=CLASS        find files of object class CLASS" << endl;
	cout << endl;

	cout << "OPTIONS:" << endl;
	cout << "  -d, --direct                   do not search for type's attributes" << endl;
	cout << "  -R, --regex                    enable regular expressions" << endl;
	cout << "  -h, --help                     print this help text and exit" << endl;
	cout << "  -V, --version                  print version information and exit" << endl;
	cout << endl;
	cout << "If the fclist does not contain MLS ranges and -m was given," << endl;
	cout << "then the search will return nothing." << endl;
}

int main(int argc, char *argv[])
{
	int optc;
	sefs_query *query = new sefs_query();

	char *type = NULL;
	bool indirect = true;
	try
	{
		while ((optc = getopt_long(argc, argv, "t:u:r:m:p:c:RhV", longopts, NULL)) != -1)
		{
			switch (optc)
			{
			case 't':
				type = optarg;
				break;
			case 'u':
				query->user(optarg);
				break;
			case 'r':
				query->role(optarg);
				break;
			case 'm':
				query->range(optarg, APOL_QUERY_EXACT);
				break;
			case 'p':
				query->path(optarg);
				break;
			case 'c':
				query->objectClass(optarg);
				break;
			case 'd':
				indirect = false;
				break;
			case 'R':
				query->regex(true);
				break;
			case 'h':     // help
				usage(argv[0], false);
				exit(0);
			case 'V':     // version
				cout << "findcon " << VERSION << endl << COPYRIGHT_INFO << endl;
				exit(0);
			default:
				usage(argv[0], true);
				exit(1);
			}
		}
		query->type(type, indirect);
	}
	catch(bad_alloc)
	{
		delete query;
		exit(-1);
	}
	if (optind + 1 != argc)
	{
		usage(argv[0], 1);
		delete query;
		exit(-1);
	}

	// try to autodetect the type of thing being searched
	struct stat sb;
	if (stat(argv[optind], &sb) != 0)
	{
		cerr << "Could not open " << argv[optind] << ": " << strerror(errno) << endl;
		delete query;
		exit(-1);
	}

	sefs_fclist *fclist = NULL;
	apol_vector_t *results = NULL;
	try
	{
		if (S_ISDIR(sb.st_mode))
		{
			fclist = new sefs_filesystem(argv[optind], NULL, NULL);
		}
		else if (sefs_db::isDB(argv[optind]))
		{
			fclist = new sefs_db(argv[optind], NULL, NULL);
		}
		else
		{
			fclist = new sefs_fcfile(argv[optind], NULL, NULL);
		}

		results = fclist->runQuery(query);
		for (size_t i = 0; i < apol_vector_get_size(results); i++)
		{
			const sefs_entry *e = static_cast < sefs_entry * >(apol_vector_get_element(results, i));
			char *str = e->toString();
			cout << str << endl;
			free(str);
		}
	}
	catch(...)
	{
		delete query;
		delete fclist;
		apol_vector_destroy(&results);
		exit(-1);
	}

	delete query;
	delete fclist;
	apol_vector_destroy(&results);
	return 0;
}
